### PR TITLE
feat!: add the ability to invoke an API to contract governance

### PR DESCRIPTION
**Pull request by Chris-Hibbert**
refs: #4188

## Description

Add to contract governance the ability to invoke particular declared methods under governance..

This is parallel to parameter governance, and uses similar mechanisms and legibility. The set of APIs that are eligible to be invoked this way are declared by the contract by listing them in an `getGovernedApis` method in the creator facet, and listing them in terms.

### Security Considerations

For legibility, clients of the contract have to be able to tell which methods can be called under governance, and they have to be able to examine the election manager to see how they can be invoked.

### Documentation Considerations

This needs to be documented along with the rest of governance.

### Testing Considerations

The existing test of contract governance should show that governance can invoke this.

### CODE REVIEW

**Comment by Chris-Hibbert** [comment]
@turadg   I'm hoping you'll be willing to review the governance aspects of this PR. Otherwise I may have to call in @erights.  At a minimum, I'd like a review (and suggestions) for the TypeScript declarations.

File: packages/governance/src/question.js [hunk 1]
```diff
@@ -80,6 +83,14 @@ const looksLikeParamChangeIssue = issue => {
   assertInstance(issue.contract);
 };
 
+const looksLikeApiInvocation = issue => {
```
Comment (turadg) [inline packages/governance/src/question.js]: The type annotations in the switch aren't doing what I suspect they were meant to. I think you're after  [Assertion Functions](https://devblogs.microsoft.com/typescript/announcing-typescript-3-7/#assertion-functions) .

```suggestion
/**
 * @param {*} issue
 * @returns {asserts issue is ApiInvocationIssue}
 */
const looksLikeApiInvocation = issue => {
```

File: packages/governance/src/types.js [hunk 2]
```diff
@@ -351,6 +351,13 @@
  * @property {ParamValue} proposedValue
  */
 
+/**
+ * @typedef {Object} ApiInvocationIssue
+ * @property {string} methodName
+ * @property {Instance} contract
+ * @property {any} methodParams
```
Comment (turadg) [inline packages/governance/src/types.js]: Is this accurate?
```suggestion
 * @property {Record<string, Callable>`} methodParams
```
(`Callable` is defined in `@agoric/eventual-send`)

File: packages/governance/src/types.js [hunk 3]
```diff
@@ -490,11 +497,18 @@
...
+ */
+
+/**
+ * @template {GovernableCreatorFacet} [CF=any]
+ * @typedef {CF} LimitedCreatorFacet
  *
  * The creatorFacet for the governed contract that will be passed to the
  * responsible party. It does not have access to the paramManager.
- * @property {() => VoteOnParamChange} getContractGovernor
+ * @property {() => Instance} getContractGovernor
```
Comment (turadg) [inline packages/governance/src/types.js:483]: Oh good, I was confused about this in another PR.

File: packages/governance/src/types.js [hunk 4]
```diff
@@ -490,11 +497,18 @@
...
 
 /**
- * @typedef {Object} LimitedCreatorFacet
+ * @typedef {{
+ *   getGovernedApis: () => unknown
+ * }} GovernableCreatorFacet
+ */
+
+/**
+ * @template {GovernableCreatorFacet} [CF=any]
```
Comment (turadg) [inline packages/governance/src/types.js]: If this isn't too onerous:
```suggestion
 * @template {GovernableCreatorFacet} [CF=unknown]
```

File: packages/governance/src/types.js [hunk 5]
```diff
@@ -542,15 +557,15 @@
...
 
 /**
- * @typedef {Object} GovernedCreatorFacet
+ * @template {object} CF creator facet
+ * @typedef {CF} GovernedCreatorFacet
  * @property {() => ParamManagerRetriever} getParamMgrRetriever - allows accessing
  *   and updating governed parameters. Should only be directly accessible to the
  *   contractGovernor
- * @property {() => T} getLimitedCreatorFacet - the creator
+ * @property {() => CF & LimitedCreatorFacet} getLimitedCreatorFacet - the creator
```
Comment (turadg) [inline packages/governance/src/types.js]: `LimitedCreatorFacet` is defined as a generic type with original CF.
```suggestion
 * @property {() => LimitedCreatorFacet<CF>} getLimitedCreatorFacet - the creator
```
The union would make sense if we made the type be "additions to a governed creator facet". We've talked about the problems with that name but it's probably better to do in another PR.

File: packages/governance/src/types.js [hunk 6]
```diff
@@ -593,12 +617,24 @@
...
 
+/**
+ * @typedef {Object} ApiGovernor
+ * @property {VoteOnApiInvocation} voteOnApiInvocation
+ * @property {CreatedQuestion} createdQuestion
+ */
+
+/**
+ * @callback GetUpdatedPoserFacet
+ * @returns {Promise<PoserFacet>}
```
Comment (turadg) [inline packages/governance/src/types.js]: this isn't really a callback. just inline where this is used,
```
* @param {() => Promise<PoserFacet>} getUpdatedPoserFacet
```

File: packages/governance/test/swingsetTests/contractGovernor/bootstrap.js [hunk 7]
```diff
@@ -163,6 +163,27 @@ const setupElectorateChange = async (
...
+    'governanceApi',
+    [], // empty params
+    installations.binaryVoteCounter,
+    2n,
+  );
+
+  E(E(zoe).getPublicFacet(instance))
+    .getOutcome()
+    .then(outcome => log(`vote outcome: ${q(outcome)}`))
+    .catch(e => log(`vote failed ${e}`));
```
Comment (turadg) [inline packages/governance/test/swingsetTests/contractGovernor/bootstrap.js:179]: tangential, can/should our log distinguish between error messages and others? (likd stdout stderr)

**Review by turadg — COMMENTED** [review]
> At a minimum, I'd like a review (and suggestions) for the TypeScript declarations.

☑️

> I'm hoping you'll be willing to review the governance aspects of this PR.

Will take a complete pass later today.

File: packages/governance/src/question.js (hunk 1 above)
Comment (Chris-Hibbert) [reply packages/governance/src/question.js]: done

File: packages/governance/src/types.js (hunk 2 above)
Comment (Chris-Hibbert) [reply packages/governance/src/types.js]: I think I really meant Array<any>. What did you intend to convey with `Callable`?

 (I presume the single quote is just a typo, and not some syntax I haven't learned yet.)  

[after discussion]

I think we ended up with something that should mean an Array whose type will come from the contract. I put in `{[unknown]}`, hoping that that's how one writes a TBD tuple.

File: packages/governance/src/types.js (hunk 4 above)
Comment (Chris-Hibbert) [reply packages/governance/src/types.js]: When I do that, I get *Type 'unknown' does not satisfy the constraint 'GovernableCreatorFacet'*.

File: packages/governance/src/types.js (hunk 5 above)
Comment (Chris-Hibbert) [reply packages/governance/src/types.js]: done

File: packages/governance/src/types.js (hunk 6 above)
Comment (Chris-Hibbert) [reply packages/governance/src/types.js]: done

File: packages/governance/test/swingsetTests/contractGovernor/bootstrap.js (hunk 7 above)
Comment (Chris-Hibbert) [reply packages/governance/test/swingsetTests/contractGovernor/bootstrap.js:179]: This is the log that produces golden output. Since the usage of that is just to compare all output against a correct output, I don't think it helps to split the output, or label errors distinctly.

File: packages/governance/src/types.js [hunk 8]
```diff
@@ -205,6 +205,12 @@
  * @returns {VoteCounterFacets}
  */
 
+/**
+ * @callback LooksLikeApiInvocation
```
Comment (turadg) [inline packages/governance/src/types.js]: why name this as a type? the type looks good (bonus points for `unknown`) but it's making one more named type that doesn't seem to add any value over typing the `looksLikeApiInvocation` function.

If you want a generic interface to re-use, I'd rename to `AssertsApiInvocationIssue`, but even then fi we want that more generic functionality I'd say to pull it out into a utility module and have the type on the function.

File: packages/zoe/test/unitTests/zcf/setupZcfTest.js [hunk 9]
```diff
@@ -48,7 +48,7 @@ export const setupZCFTest = async (issuerKeywordRecord, terms) => {
   ).startInstance(installation, issuerKeywordRecord, terms);
   const { vatAdminState } = fakeVatAdmin;
   // @ts-ignore fix types to understand that zcf is always defined
-  assert(zcf !== undefined);
+  assert(zcf !== undefined, 'zcf is required');
```
Comment (turadg) [inline packages/zoe/test/unitTests/zcf/setupZcfTest.js]: We can help the tester a little more here. Also it's idiomatic to check for presence with a truthiness.

```suggestion
  // @ts-expect-error setZCF may not have been called yet
  assert(zcf, 'zcf is required; did you forget to setZCF?');
```

Comment (Chris-Hibbert) [reply packages/zoe/test/unitTests/zcf/setupZcfTest.js]: done

File: packages/governance/src/types.js (hunk 8 above)
Comment (Chris-Hibbert) [reply packages/governance/src/types.js]: thanks for your (out-of-band) help fixing the delcarations.

**Comment by Tartuffo** [comment]
@Chris-Hibbert is this strictly needed for MN-1 launch?

**Comment by Chris-Hibbert** [comment]
> is this strictly needed for MN-1 launch?

Yes, according to @rowgraus, @dtribble, and @btulloh

File: packages/governance/src/types.js (hunk 8 above)
Comment (erights) [reply packages/governance/src/types.js]: > thanks for your (out-of-band) help fixing the delcarations.

Does the current PR reflect this out-of-band effort yet?

Comment (Chris-Hibbert) [reply packages/governance/src/types.js]: Yep. The typescript declarations reflect the current state of the code and make a lot of the API visible in an IDE.

File: packages/governance/README.md [hunk 10]
```diff
@@ -70,32 +70,37 @@ voterHandle.
...
-
-The governor starts up the Contract and can see what parameters are subject to
-governance. It provides a private facet that carries the ability to request
-votes on changing particular parameters. Some day we may figure out how to make
-the process and schedule of selecting parameter changes to vote on also subject
-to governance, but that's too many meta-levels at this point.
+We want some contracts to be able to make the manner of their governance
+legible, that is that changes to their parameters, and the ability to invoke
+some management methods are controlled by a public process. Observers (clients
+and voters) must be able to see who has control, and how those changes can
```
Comment (turadg) [inline packages/governance/README.md]: Glad to see this definition of this important jargon. Consider for "governed":
> Contracts can provide aspects that are _governed_ by a public process.
> Specifically,
> - changes to a defined subset of parameters
> - invocation of a defined subsets of methods

(Should we name those two aspects of governance?)

Then "legibility":

> The governance package enforces _legibility_, meaning that observers 
> (clients and voters) are able to see who has control and how those changes
> can happen. To make control of parameters legible…

File: packages/governance/README.md [hunk 11]
```diff
@@ -70,32 +70,37 @@ voterHandle.
...
+request votes on changing parameters and invoking the controlled APIs. At some
+point, we may add governance control over the process and schedule of calling
+those votes, but that hasn't been planned in detail yet.
 
 The party that has the question-creating facet of the ContractGovernor can
-create a question that asks about changing a particular parameter on the
-contract instance. The electorate object creates new questions, and makes a new
-instance of a VoteCounter so everyone can see how questions will be counted.
-
+create questions on parameters or APIs for that contract instance. The
```
Comment (turadg) [inline packages/governance/README.md:101]: Out of scope and just thinking out loud: "Questions" sounds like a query for information. Questions can be answered in many ways, but these questions get resolved with exactly one answer. I suppose that's a fact of how the question is used rather than the question itself. "Resolution" would be too specific because some questions actually are queries for information (e.g. SURVEY)

File: packages/governance/README.md [hunk 12]
```diff
@@ -118,8 +123,9 @@ provided by the governed contract and doesn't include any powerful governance
 capabilities. ContractGovernor starts the governed contract, so it gets the
 powerful creatorFacet. ContractGovernor needs access to the paramManager, but
 shouldn't share it. So the contractGovernor's creatorFacet provides access to
-the governed contract's publicFacet, creatorFacet, instance, and
-`voteOnParamChange()`, which the contract's owner should treat as powerful. 
+the governed contract's publicFacet, creatorFacet, instance,
```
Comment (turadg) [inline packages/governance/README.md]: If `publicFacet` is meant to be indicate the part of the response type then it should be backticked. If not, it should be the words "public facet". I figure it's the former since the `voteOnApiInvocation()` is backticked. Though I'm not sure it needs ()` until it's being invoked :)

File: packages/governance/README.md [hunk 13]
```diff
@@ -118,8 +123,9 @@ provided by the governed contract and doesn't include any powerful governance
...
+`voteOnApiInvocation()`, and `voteOnParamChange()`, which the contract's owner
```
Comment (turadg) [inline packages/governance/README.md]: I can't tell which items the "which" applies to.

File: packages/governance/README.md [hunk 14]
```diff
@@ -198,6 +212,21 @@ uses the values passed in `terms` to create the paramManager, reviewers of the
...
+```
+  terms: {
+    main: {
+      [MALLEABLE_NUMBER]: { type: ParamTypes.NAT, value: number },
+      [CONTRACT_ELECTORATE]: {
+        type: ParamTypes.INVITATION,
+        value: invitationAmount,
+      },
+    },
+    governedApis: ['governanceApi'],
```
Comment (turadg) [inline packages/governance/README.md]: `governanceApi` sounds like a term of art. consider "doThingChosenByElectorate"

File: packages/governance/README.md [hunk 15]
```diff
@@ -198,6 +212,21 @@ uses the values passed in `terms` to create the paramManager, reviewers of the
 contract can verify that all and only the declared parameters are under the
 control of the paramManager and made visible to the contract's clients.
 
+Governed methods and parameters must be included in terms.
+
+```
+  terms: {
+    main: {
+      [MALLEABLE_NUMBER]: { type: ParamTypes.NAT, value: number },
```
Comment (turadg) [inline packages/governance/README.md:225]: out of scope: WDYT of changing to a tuple?
```
[MALLEABLE_NUMBER]: [ParamTypes.NAT, number],
```
Do we anticipate other elements?

File: packages/governance/src/contractGovernance/governApi.js [hunk 16]
```diff
@@ -0,0 +1,132 @@
...
+  ElectionType,
+  coerceQuestionSpec,
+} from '../question.js';
+
+const { details: X } = assert;
+
+/**
+ * Make a pair of positions for a question about whether to invoke an API. If
+ * the vote passes, the method will be called on the governedApis facet with the
+ * parameters that were provided.
```
Comment (turadg) [inline packages/governance/src/contractGovernance/governApi.js]: isn't this the _arguments_ provided? These are the actual values to be used in the call (which will be bound in the method body to its parameters)

File: packages/governance/src/contractGovernance/governApi.js [hunk 17]
```diff
@@ -0,0 +1,132 @@
...
+    // * If the vote was negative, resolve to the outcome
+    // * If the API invocation succeeds, say so
+    // * If the update fails, reject the promise
+    // * if the vote outcome failed, reject the promise.
+    E(counterPublicFacet)
+      .getOutcome()
+      .then(outcome => {
+        if (keyEQ(positive, outcome)) {
+          E(governedApis)
+            [apiMethodName](...methodParams)
```
Comment (turadg) [inline packages/governance/src/contractGovernance/governApi.js]: What is to be invoked is defined in multiple places:
- `positive`
- `outcome`
- `issue`
- local scope from parameters of `voteOnApiInvocation`

I see a check that `outcome` and `positive` match, because that's a signal as to what the outcome was.

As an extra security check, and to make clear to the reader, please check here that the outcome matches the parameters. E.g.,
```suggestion
        if (keyEQ(positive, outcome)) {
          keyEQ(outcome, { apiMethodName, methodParams });
          E(governedApis)
            [apiMethodName](...methodParams)
```

Though as noted above, I think this must be `methodArgs`.

File: packages/governance/src/contractGovernance/governApi.js [hunk 18]
```diff
@@ -0,0 +1,132 @@
...
+  const negative = harden({ dontInvoke: apiMethodName });
+  return { positive, negative };
+};
+
+/**
+ * manage contracts that allow governance to invoke functions.
+ *
+ * @param {ERef<ZoeService>} zoe
+ * @param {Instance} governedInstance
+ * @param {any} governedApis
```
Comment (turadg) [inline packages/governance/src/contractGovernance/governApi.js]:
```suggestion
 * @param {ERef<{ [methodName: string]: (...args: any) => unknown }>} governedApis
```

File: packages/governance/src/contractGovernance/governApi.js [hunk 19]
```diff
@@ -0,0 +1,132 @@
...
+
+    const { positive, negative } = makeApiInvocationPositions(
+      apiMethodName,
+      methodParams,
+    );
+
+    /** @type {ApiInvocationIssue} */
+    const issue = harden({ apiMethodName, methodParams });
+    const questionSpec = coerceQuestionSpec({
+      method: ChoiceMethod.UNRANKED,
```
Comment (turadg) [inline packages/governance/src/contractGovernance/governApi.js:72]: Looking up what `UNRANKED` means,
>  * "unranked" is more formally known as "approval" voting, but this is hard for
>  * people to intuit when there are only two alternatives.

Out of scope but I question this conclusion. "unranked" suggest to me "not ranked voting" which suggests the most common type of voting: pick one. I contend `ChoiceMethod.APPROVE` would be more clear. "Pick what you approve"

File: packages/governance/src/contractGovernance/governApi.js [hunk 20]
```diff
@@ -0,0 +1,132 @@
...
+    const { publicFacet: counterPublicFacet, instance: voteCounter } = await E(
+      getUpdatedPoserFacet(),
+    ).addQuestion(voteCounterInstallation, questionSpec);
+
+    voteCounters.add(voteCounter);
+
+    // CRUCIAL: Here we wait for the voteCounter to declare an outcome, and then
+    // attempt to invoke the API if that's what the vote called for. We need to
+    // make sure that outcomeOfUpdateP is updated whatever happens.
+    //
```
Comment (turadg) [inline packages/governance/src/contractGovernance/governApi.js:91]: what if it gets closed by exceeding the deadline? resolved negative or rejected?

File: packages/governance/src/contractGovernance/governApi.js [hunk 21]
```diff
@@ -0,0 +1,132 @@
...
+    // * If the update fails, reject the promise
+    // * if the vote outcome failed, reject the promise.
+    E(counterPublicFacet)
+      .getOutcome()
+      .then(outcome => {
+        if (keyEQ(positive, outcome)) {
+          E(governedApis)
+            [apiMethodName](...methodParams)
+            .then(returnValue => outcomeOfUpdateP.resolve(returnValue))
+            .catch(e => {
```
Comment (turadg) [inline packages/governance/src/contractGovernance/governApi.js]: Four cases defined here. If we use regular promises instead of promisekit, a throw will automatically reject. So think think this can be simplified to:
```
    const outcomeOfUpdate = E(counterPublicFacet)
      .getOutcome()
      .then(outcome => {
        if (keyEQ(positive, outcome)) {
          return E(governedApis)[apiMethodName](...methodParams);
        } else {
          return negative;
        }
      });
```
Now it's much easier for me to see that it resolves to either the return value of the method call or an object representing the position of a negative outcome. This is confusing. The term "outcome" is used here to mean "positive or negative" but "outcomeOfUpdate" returns "negative or the return value from executing the positive outcome".

What are the use cases for getting a result back from the method call? Since it's something people vote on executing, it seems like you would only execute "void" methods. In which case all the caller needs to know is a) whether the vote went positive or negative and b) that when they get this answer that if it was positive then that call is completed. IIUC, alll the other states are exceptional (e.g. failure to make remote call)

File: packages/governance/src/contractHelper.js [hunk 22]
```diff
@@ -68,9 +68,9 @@ const handleParamGovernance = (zcf, paramManager) => {
   };
 
   /**
-   * @template CF
-   * @param {CF} originalCreatorFacet
-   * @returns {CF & LimitedCreatorFacet}
+   * @template {GovernableCreatorFacet} CF creator facet
```
Comment (turadg) [inline packages/governance/src/contractHelper.js]: 👍

File: packages/governance/src/contractHelper.js [hunk 23]
```diff
@@ -80,20 +80,24 @@ const handleParamGovernance = (zcf, paramManager) => {
...
     // exclusively for contractGovernor, which only reveals limitedCreatorFacet
     return Far('creatorFacet', {
+      // @ts-expect-error special case
       getParamMgrRetriever: () => {
         return Far('paramRetriever', { get: () => paramManager });
       },
       getInvitation: name => paramManager.getInternalParamValue(name),
       getLimitedCreatorFacet: () => limitedCreatorFacet,
+      getGovernedApis: () => Far('governedAPIs', governedApis),
+      getGovernedApiNames: () => Object.keys(governedApis),
```
Comment (turadg) [inline packages/governance/src/contractHelper.js:105]: We talked about getting the API names from the governed API remote call. IIUC we can't because the whole object is Far and there's no way to make it an copyable object of method names mapped to Far methods. Is that right?

I request that be documented as explanation for the surprise that the API requires a `getGovernedApiNames` that is seemingly redundant with getting the keys of what's returned by `getGovernedApis`.

File: packages/governance/src/question.js [hunk 24]
```diff
@@ -33,6 +33,7 @@ const ChoiceMethod = {
  *   PARAM_CHANGE: 'param_change',
  *   ELECTION: 'election',
  *   SURVEY: 'survey',
+ *   API_INVOCATION: 'api_invocation',
```
Comment (turadg) [inline packages/governance/src/question.js]: we can DRY this out by replacing this type annotation with,
```
const ElectionType = /** @type {const} */ {
```

File: packages/governance/src/question.js [hunk 25]
```diff
@@ -80,6 +83,15 @@ const looksLikeParamChangeIssue = issue => {
   assertInstance(issue.contract);
 };
 
+/** @type {LooksLikeApiInvocation} */
```
Comment (turadg) [inline packages/governance/src/question.js]: please inline this type since it's only purpose is to annotate this function.

File: packages/governance/src/question.js [hunk 26]
```diff
@@ -80,6 +83,15 @@ const looksLikeParamChangeIssue = issue => {
...
+const looksLikeApiInvocation = issue => {
```
Comment (turadg) [inline packages/governance/src/question.js]: We already have a word for "looks like":
```suggestion
const assertApiInvocation = issue => {
```

I think it would be good to revise the untouched `looksLike` functions too.

File: packages/governance/src/types.js [hunk 27]
```diff
@@ -583,8 +606,17 @@
...
- * @param {bigint} deadline
- * @returns {ParamChangeVoteResult}
+ * @param {Timestamp} deadline
+ * @returns {ContractGovernanceVoteResult}
+ */
+
+/**
+ * @callback VoteOnApiInvocation
+ * @param {string} apiMethodName
+ * @param {[unknown]} methodParams
```
Comment (turadg) [inline packages/governance/src/types.js]: brackets with contents indicate a tuple. I think this is an array (single type for all elements and unknown length).
```suggestion
 * @param {unknown[]} methodArgs
```

File: packages/governance/README.md (hunk 11 above)
Comment (Chris-Hibbert) [reply packages/governance/README.md:101]: To the extent the elections field has a term that encompasses both referendums and elections to office I think it's "ballot question". "ballot measures" usually seek to enact a law. I see "question" more often for initiatives, recalls, advisory measures than for elections, but people understand in the latter case that they are answering the question "who shall be selected to fill this office?"

File: packages/governance/README.md (hunk 12 above)
Comment (Chris-Hibbert) [reply packages/governance/README.md]: done

File: packages/governance/README.md (hunk 13 above)
Comment (Chris-Hibbert) [reply packages/governance/README.md]: clarified.

File: packages/governance/README.md (hunk 15 above)
Comment (Chris-Hibbert) [reply packages/governance/README.md:225]: I wouldn't have a problem with that. Would it make it harder to validate (because of indeterminate order?)  main: is for governed params. If we add more kinds, then, like governedApis, they'd be placed in parallel sections.

File: packages/governance/src/contractGovernance/governApi.js (hunk 16 above)
Comment (Chris-Hibbert) [reply packages/governance/src/contractGovernance/governApi.js]: fixed

File: packages/governance/src/contractGovernance/governApi.js (hunk 18 above)
Comment (Chris-Hibbert) [reply packages/governance/src/contractGovernance/governApi.js]: thanks

File: packages/governance/src/contractGovernance/governApi.js (hunk 19 above)
Comment (Chris-Hibbert) [reply packages/governance/src/contractGovernance/governApi.js:72]: I would contend that approval voting is one kind of unranked voting. The normal case for approval voting is not to limit the number of votes any voter can cast. I've seen elections for school board or other multiple-seat elections where voters are instructed to "vote for up to N", where N is usually the number of seats to be filled. I'm calling this "unranked", since it's not classical approval voting.

In approval voting, you can vote for three candidates in a one-winner race. The idea is to encourage more voters to indicate all the candidates they would find acceptable so we can find the intersection of interest groups.

You're probably right that voters wouldn't be confused by calling it APPROVE, but it doesn't seem to me to reflect voting terminology.

File: packages/governance/src/contractGovernance/governApi.js (hunk 20 above)
Comment (Chris-Hibbert) [reply packages/governance/src/contractGovernance/governApi.js:91]: We don't yet have elections that can close before the deadline. (I think we want them for emergency actions.) 

This code doesn't attempt to decide whether the ballot counter missed its deadline.

File: packages/governance/src/contractGovernance/governApi.js (hunk 17 above)
Comment (Chris-Hibbert) [reply packages/governance/src/contractGovernance/governApi.js]: Ahh!   I didn't catch that the comment on documentation also applied to the variable name in the code. I've changed the name and added the assertion.

File: packages/governance/src/contractGovernance/governApi.js (hunk 21 above)
Comment (Chris-Hibbert) [reply packages/governance/src/contractGovernance/governApi.js]: Thanks for the suggestion. I simplified in both ways (dropping explicit promises, and returning the positive position on success.)

File: packages/governance/src/question.js (hunk 25 above)
Comment (Chris-Hibbert) [reply packages/governance/src/question.js]: done.

File: packages/governance/src/question.js (hunk 24 above)
Comment (Chris-Hibbert) [reply packages/governance/src/question.js]: done after additional consultation. It ended up needing
```
const ElectionType = /** @type {const} */ ({
```
with parens around the object.

File: packages/governance/README.md (hunk 10 above)
Comment (Chris-Hibbert) [reply packages/governance/README.md]: done

File: packages/governance/src/contractHelper.js (hunk 23 above)
Comment (Chris-Hibbert) [reply packages/governance/src/contractHelper.js:105]: added.

File: packages/governance/src/question.js (hunk 26 above)
Comment (Chris-Hibbert) [reply packages/governance/src/question.js]: I think we use "looksLike" when the verification is leaky.  (`looksLikeBrand`)  But I won't say that any of these are leaky or unreliable in the same way, so I've changed them.

File: packages/governance/README.md [hunk 28]
```diff
@@ -198,6 +217,21 @@ uses the values passed in `terms` to create the paramManager, reviewers of the
...
+      [CONTRACT_ELECTORATE]: {
+        type: ParamTypes.INVITATION,
+        value: invitationAmount,
+      },
+    },
+    governedApis: ['makeItSo'],
```
Comment (turadg) [inline packages/governance/README.md:231]: 🖖

File: packages/governance/docs/contractGovernance.puml [hunk 29]
```diff
@@ -10,7 +10,7 @@ package "GovernedContract Vat" <<Rectangle>>  {
     }
 
     class GovernedContract {
-        <i>verifiable</i>: Governor, params
+        <i>verifiable</i>: Governor, params, governedApis
```
Comment (turadg) [inline packages/governance/docs/contractGovernance.puml:13]: I looked into replacing this with Mermaid.js so it wouldn't need a build step. Unfortunately it [doesn't support package in class digrams](https://github.com/mermaid-js/mermaid/issues/1052).

File: packages/governance/src/contractGovernance/governApi.js [hunk 30]
```diff
@@ -0,0 +1,129 @@
...
+    // * If the vote passed, invoke the API, and return the positive position
+    // * If the vote was negative, return the negative position
+    // * If we can't do either, (the vote failed or the API invocation failed)
+    //   return a broken promise.
+    const outcomeOfUpdate = E(counterPublicFacet)
+      .getOutcome()
+      // @ts-expect-error return types don't appear to match
+      .then(outcome => {
+        if (keyEQ(positive, outcome)) {
+          keyEQ(outcome, harden({ apiMethodName, methodArgs }));
```
Comment (turadg) [inline packages/governance/src/contractGovernance/governApi.js]: necessary change:
```suggestion
          assert(keyEQ(outcome, harden({ apiMethodName, methodArgs })));
```

File: packages/governance/src/contractGovernor.js [hunk 31]
```diff
@@ -135,18 +136,73 @@ const start = async zcf => {
...
+        governedNames,
+        timer,
+        getUpdatedPoserFacet,
+      );
+    }
+
+    // if we aren't governing APIs, voteOnApiInvocation shouldn't be called
+    return {
+      voteOnApiInvocation: () => {
+        throw Error('api governance not configured');
```
Comment (turadg) [inline packages/governance/src/contractGovernor.js:191]: I'm not certain when to use this instead but I've seen:
```suggestion
        assert.fail('api governance not configured');
```

File: packages/governance/src/contractGovernor.js [hunk 32]
```diff
@@ -135,18 +136,73 @@ const start = async zcf => {
...
+      },
+      createdQuestion: () => false,
+    };
+  };
+
+  const { voteOnApiInvocation, createdQuestion } = await initApiGovernance();
```
Comment (turadg) [inline packages/governance/src/contractGovernor.js]: Above there's a helpful renaming destructure: `createdQuestion: createdParamQuestion`
```suggestion
  const { voteOnApiInvocation, createdQuestion: createdApiQuestion } = await initApiGovernance();
```

File: packages/governance/src/question.js [hunk 33]
```diff
@@ -78,10 +69,28 @@ const looksLikeParamChangeIssue = issue => {
   assert(issue && issue.proposedValue);
   const assertInstance = makeAssertInstance('contract');
   assertInstance(issue.contract);
+  return undefined;
```
Comment (turadg) [inline packages/governance/src/question.js]: stray, remove

File: packages/governance/src/question.js [hunk 34]
```diff
@@ -90,23 +99,31 @@ const looksLikeIssueForType = (electionType, issue) => {
...
     case ElectionType.ELECTION:
-      looksLikeSimpleIssue(/** @type {SimpleIssue} */ (issue));
+      assertSimpleIssue(/** @type {SimpleIssue} */ (issue));
       break;
     case ElectionType.PARAM_CHANGE:
-      looksLikeParamChangeIssue(/** @type {ParamChangeIssue} */ (issue));
+      assertParamChangeIssue(/** @type {ParamChangeIssue} */ (issue));
+      break;
+    case ElectionType.API_INVOCATION:
+      assertApiInvocation(/** @type {ApiInvocationIssue} */ (issue));
```
Comment (turadg) [inline packages/governance/src/question.js]: remove casts. we don't know what they really are at this point

File: packages/governance/src/question.js [hunk 35]
```diff
@@ -90,23 +99,31 @@ const looksLikeIssueForType = (electionType, issue) => {
...
       break;
     default:
       throw Error(`Election type unrecognized`);
   }
+  return undefined;
```
Comment (turadg) [inline packages/governance/src/question.js]: ditto as above.

if this is to make eslint happy, eslint has a bug in a rule which should be suppressed

File: packages/governance/src/types.js [hunk 36]
```diff
@@ -351,6 +319,12 @@
  * @property {ParamValue} proposedValue
  */
 
+/**
+ * @typedef {Object} ApiInvocationIssue
+ * @property {string} apiMethodName
+ * @property {[unknown]} methodArgs
```
Comment (turadg) [inline packages/governance/src/types.js]: array, not 1-tuple
```suggestion
 * @property {unknown[]} methodArgs
```

File: packages/governance/src/types.js [hunk 37]
```diff
@@ -450,6 +424,17 @@
  * @property {ParamValue} proposedValue
  */
 
+/**
+ * @typedef {Object} InvokeApiPosition
+ * @property {string} apiMethodName
+ * @property {any} methodArgs
```
Comment (turadg) [inline packages/governance/src/types.js]:
```suggestion
 * @property {unknown[]} methodArgs
```
or `any[]` if that makes a lot of errors. but we at least know that it's an array

File: packages/governance/src/types.js [hunk 38]
```diff
@@ -490,11 +475,17 @@
  */
 
 /**
- * @typedef {Object} LimitedCreatorFacet
+ * @typedef {{
+ * }} GovernableCreatorFacet
```
Comment (turadg) [inline packages/governance/src/types.js]: I forget, is there really no requirement on a creator facet to be governed?

if so, why name it?

File: packages/governance/src/types.js [hunk 39]
```diff
@@ -583,8 +575,17 @@
...
- * @param {bigint} deadline
- * @returns {ParamChangeVoteResult}
+ * @param {Timestamp} deadline
+ * @returns {ContractGovernanceVoteResult}
+ */
+
+/**
+ * @callback VoteOnApiInvocation
+ * @param {string} apiMethodName
+ * @param {[unknown]} methodArgs
```
Comment (turadg) [inline packages/governance/src/types.js]:
```suggestion
 * @param {unknown[]} methodArgs
```

File: packages/run-protocol/src/vaultFactory/vaultFactory.js [hunk 40]
```diff
@@ -240,6 +240,8 @@ export const start = async (zcf, privateArgs) => {
     getParamMgrRetriever,
     getInvitation: electorateParamManager.getInternalParamValue,
     getLimitedCreatorFacet: () => vaultFactory,
+    getGovernedApis: () => harden({}),
+    getGovernedApiNames: () => harden({}),
```
Comment (turadg) [inline packages/run-protocol/src/vaultFactory/vaultFactory.js:278]: I thought we said that undefined getters would be treated the same as getters returning empty. This is busywork for the contract author.

**Review by turadg — COMMENTED** [review]
Very close!

File: packages/governance/src/contractGovernance/governApi.js (hunk 30 above)
Comment (Chris-Hibbert) [reply packages/governance/src/contractGovernance/governApi.js]: good catch.

File: packages/governance/src/contractGovernor.js (hunk 31 above)
Comment (Chris-Hibbert) [reply packages/governance/src/contractGovernor.js:191]: `throw Error` is much more common in contracts.  I'm not sure that `assert.fail` doesn't work as well, but I know that the former works well with our distributed stack trace utilities.

File: packages/governance/src/contractGovernor.js (hunk 32 above)
Comment (Chris-Hibbert) [reply packages/governance/src/contractGovernor.js]: good call.

File: packages/governance/src/question.js (hunk 33 above)
Comment (Chris-Hibbert) [reply packages/governance/src/question.js]: It makes the `returns ... assert` not need an `@typescript-ignore`.

File: packages/governance/src/question.js (hunk 35 above)
Comment (Chris-Hibbert) [reply packages/governance/src/question.js]: That's what it's for. I thought the return was less ugly than the `ignore`.  If your claim is that the broken rule should be called out, I'll change it back.

File: packages/governance/src/question.js (hunk 34 above)
Comment (Chris-Hibbert) [reply packages/governance/src/question.js]: done

File: packages/governance/src/types.js (hunk 36 above)
Comment (Chris-Hibbert) [reply packages/governance/src/types.js]: done

File: packages/governance/src/types.js (hunk 37 above)
Comment (Chris-Hibbert) [reply packages/governance/src/types.js]: done

File: packages/governance/src/types.js (hunk 38 above)
Comment (Chris-Hibbert) [reply packages/governance/src/types.js]: Everything required is provided. I'll drop it.  I don't think we'll be adding requirements back.

File: packages/governance/src/types.js (hunk 39 above)
Comment (Chris-Hibbert) [reply packages/governance/src/types.js]: done

File: packages/run-protocol/src/vaultFactory/vaultFactory.js (hunk 40 above)
Comment (turadg) [reply packages/run-protocol/src/vaultFactory/vaultFactory.js:278]: This was clarified in chat: the auto-providing is done by `handleParamGovernance` which vaultFactory eschews (because it's an uncommonly complex governance)

**Review by turadg — APPROVED** [review]
👍
